### PR TITLE
[MLIR][Presburger] Make IntegerRelation::convertVarKind consistent wi…

### PR DESCRIPTION
**Pull request by iambrj**
…th PresburgerSpace::convertVarKind

### CODE REVIEW

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 1]
```diff
@@ -2260,8 +2273,7 @@ void IntegerRelation::intersectRange(const IntegerPolyhedron &poly) {
 
 void IntegerRelation::inverse() {
   unsigned numRangeVars = getNumVarKind(VarKind::Range);
-  convertVarKind(VarKind::Domain, 0, getVarKindEnd(VarKind::Domain),
-                 VarKind::Range);
+  convertVarKind(VarKind::Domain, 0, getNumDomainVars(), VarKind::Range);
```
Comment (Groverkss) [inline mlir/lib/Analysis/Presburger/IntegerRelation.cpp:2276]: nit: I prefer getNumVarKind

**Review by Groverkss — COMMENTED** [review]
Please keep this patch as something that changes API only. We can change the algorithm in a separate patch.

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 2]
```diff
@@ -1317,31 +1318,43 @@ void IntegerRelation::removeRedundantLocalVars() {
...
+  assert(dstPos <= getNumVarKind(dstKind) &&
+         "invalid position for destination variables");
 
-  if (varStart >= varLimit)
+  if (srcPos >= varLimit)
     return;
 
+  // Save the space as the insert/delete vars operations affect the identifier
+  // information in the space.
+  PresburgerSpace oldSpace = space;
```
[the remark this answers is not in the document]
Comment (iambrj) [reply mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1336]: I'm not sure how else to preserve the identifier information in the space as the IntegerRelation::convertVarKind algorithm inserts and deletes vars without preserving the identifier information. How do you suggest preserving the identifier information? Also, this patch does not change any of the insert/delete var function calls, so I'm not sure if the patch qualifies as changing the IntegerRelation::convertVarKind algorithm.

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 3]
```diff
@@ -1317,31 +1318,43 @@ void IntegerRelation::removeRedundantLocalVars() {
...
+
   // Append new local variables corresponding to the dimensions to be converted.
-  unsigned convertCount = varLimit - varStart;
-  unsigned newVarsBegin = insertVar(dstKind, pos, convertCount);
+  unsigned newVarsBegin = insertVar(dstKind, dstPos, num);
 
   // Swap the new local variables with dimensions.
   //
-  // Essentially, this moves the information corresponding to the specified ids
+  // Essentially, this moves the constraints corresponding to the specified ids
```
[the remark this answers is not in the document]
Comment (iambrj) [reply mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1343]: Okay.

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Groverkss) [reply mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1336]: Do you think you can send a patch that fixes the Identifier preservation and then send this patch? Ideally, this patch should just be a API change.

[the remark this answers is not in the document]
Comment (iambrj) [reply mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1336]: Okay, sounds good. Will send a patch implementing identifier preservation.

**Pull request closed without merging**